server/query/QueryService.java

```diff
@@ -98,95 +102,66 @@ public void execute(Transaction.Req request) {
         }
     }
 
-    private static TransactionProto.Transaction.Res response(Transaction.Req request, QueryProto.Query.Res.Builder response) {
-        return TransactionProto.Transaction.Res.newBuilder().setId(request.getId()).setQueryRes(response).build();
-    }
-
-    private void define(String queryStr, Options.Query options, Transaction.Req request) {
+    private void define(String queryStr, Options.Query options, TransactionProto.Transaction.Req req) {
         GraqlDefine query = Graql.parseQuery(queryStr).asDefine();
         Context.Query context = new Context.Query(transactionSrv.context(), options.query(query), query);
         queryMgr.define(query, context);
-        transactionSrv.respond(response(request, QueryProto.Query.Res.newBuilder()
-                .setDefineRes(QueryProto.Query.Define.Res.getDefaultInstance())));
+        transactionSrv.respond(defineRes(req.getReqId()));
     }
 
-    private void undefine(String queryStr, Options.Query options, Transaction.Req request) {
+    private void undefine(String queryStr, Options.Query options, TransactionProto.Transaction.Req req) {
         GraqlUndefine query = Graql.parseQuery(queryStr).asUndefine();
         Context.Query context = new Context.Query(transactionSrv.context(), options.query(query), query);
         queryMgr.undefine(query, context);
-        transactionSrv.respond(response(request, QueryProto.Query.Res.newBuilder()
-                .setUndefineRes(QueryProto.Query.Undefine.Res.getDefaultInstance())));
+        transactionSrv.respond(undefineRes(req.getReqId()));
     }
 
-    private void match(String queryStr, Options.Query options, Transaction.Req request) {
+    private void match(String queryStr, Options.Query options, TransactionProto.Transaction.Req eq) {
         GraqlMatch query = Graql.parseQuery(queryStr).asMatch();
         Context.Query context = new Context.Query(transactionSrv.context(), options.query(query), query);
         FunctionalIterator<ConceptMap> answers = queryMgr.match(query, context);
-        transactionSrv.stream(answers, request.getId(), context.options(), as -> response(
-                request, QueryProto.Query.Res.newBuilder().setMatchRes(
-                        QueryProto.Query.Match.Res.newBuilder().addAllAnswers(
-                                iterate(as).map(ResponseBuilder.Answer::conceptMap).toList()))
-        ));
+        transactionSrv.stream(answers, eq.getReqId(), context.options(), a -> matchResPart(eq.getReqId(), a));
```

Comment:
Should we refactor `TransactionService.stream` to not take in a lambda and instead use `map` or `flatMap`?

Reply:
I tried - it's not possible because we have to consume the iterator based on a time window first, before applying the lambda to create a "ResPart".